### PR TITLE
refactor(storage): avoid decode and compression when refill

### PR DESCRIPTION
**Pull request by MrCroxx**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Avoid redundant decode and compression when refill fetch.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/storage/src/hummock/file_cache/store.rs [hunk 1]
```diff
@@ -336,15 +351,95 @@ impl Key for SstableBlockIndex {
...
+            0 => {
+                let data = Bytes::copy_from_slice(buf);
+                let block = Block::decode_from_raw(data);
+                let block = Box::new(block);
+                Self::Loaded { block }
+            }
+            1 => {
+                let uncompressed_capacity = buf.get_u64() as usize;
+                let data = Bytes::copy_from_slice(buf);
+                let block = Block::decode(data, uncompressed_capacity)?;
```
Comment (hzxa21) [inline src/storage/src/hummock/file_cache/store.rs:436]: Will foyer update the value in place after the `read` call? Will this cause the same block being decompressed for multiple times if it is hit frequently?

Comment (MrCroxx) [reply src/storage/src/hummock/file_cache/store.rs:436]: No, after the file cache hit, the block will be populated to memory cache.

**Review by hzxa21 — APPROVED** [review]

File: src/storage/src/hummock/event_handler/refiller.rs [hunk 2]
```diff
@@ -382,14 +384,15 @@ impl CacheRefillTask {
...
+                let (range, _uncompressed_capacity) = sst.calculate_block_info(block_index);
                 let key = SstableBlockIndex {
                     sst_id: object_id,
                     block_idx: block_index as u64,
                 };
+                // see `CachedBlock::serialized_len()`
                 let mut writer = sstable_store
                     .data_file_cache()
-                    .writer(key, key.serialized_len() + uncompressed_capacity);
+                    .writer(key, key.serialized_len() + 1 + 8 + range.size().unwrap());
```
Comment (stdrc) [inline src/storage/src/hummock/event_handler/refiller.rs:395]: Could you add some comment to explain the magic number?

Comment (MrCroxx) [reply src/storage/src/hummock/event_handler/refiller.rs:395]: I've left `// see `CachedBlock::serialized_len()` for others to refer to.

**Pull request merged**